### ISSUE TITLE
Add SCSS support with Gulp and Browsersync for better UI development

### ISSUE THREAD

**Issue by prikeshsavla** [jointwt/twtxt#311]
Add Styles to the system right now is awful. Its 3 to 4 css files and the package doesnt even use a css preprocessor. 
Also with scss we can limit the styles from Pico.css and Also add our own variables and custom component styles

I would suggest using a gulp file for building css from scss and also have browsersync to add auto reload. 
This keeps the dev easier for UI and doesnt affect the overall js or package size.
I would also add a yarn/npm command to run both the dev server and gulp watcher
With gulp we can also watch for *.go files with the usual html

I want to be assigned to this @prologic

**Comment by prologic** [comment]
> Add Styles to the system right now is awful. Its 3 to 4 css files and the package doesnt even use a css preprocessor.

What's awful about it? 🤣 I _like_ to keep thing simple.

I'm not a UI/UX developer (_at all!_) and am quite terrible at CSS, but honestly SCSS doesn't really bring a lot to the table IHMO where CSS is more obvious and simpler to maintain. For large projects I guess SCSS makes sense?

> Also with scss we can limit the styles from Pico.css and Also add our own variables and custom component styles

To be honest I'd rather we create a new repo based off of @lucaslarroche's original work in [twtxt-pico](https://github.com/lucaslarroche/twtxt-pico). This should be maintained as a separate project/repo, once the SCSS is ready, compiled, released and imported into [jointwt/twtxt](https://github.com/jointwt/twtxt).
 
> I would suggest using a gulp file for building css from scss and also have browsersync to add auto reload.
> This keeps the dev easier for UI and doesnt affect the overall js or package size.
> I would also add a yarn/npm command to run both the dev server and gulp watcher
> With gulp we can also watch for *.go files with the usual html

I _really_ don't want to drag in half of Node's ecosystem into this project if I can help it. Right now the CSS/JS bundles are built with simple UNIX and Go tools and it all works very nicely.

The only thing I want to change in the near future is to migrate from [go.rice](https://github.com/GeertJohan/go.rice) to [embed](https://pkg.go.dev/embed)

> I want to be assigned to this @prologic

Okay 👍 Let's define the scope of what we _should_ be doing here. My initial thoughts are:

- Get @lucaslarroche's work from [twtxt-pico](https://github.com/lucaslarroche/twtxt-pico) into a new repo in the @jointwt org.
- Migrate existing templates to use the new styles.

**Comment by prikeshsavla** [comment]
> What's awful about it? 🤣 I like to keep thing simple.
> 
> I'm not a UI/UX developer (at all!) and am quite terrible at CSS, but honestly SCSS doesn't really bring a lot to the table IHMO where CSS is more obvious and simpler to maintain. For large projects I guess SCSS makes sense?

I understand that you are not a UI/UX dev. Scss brings a lot if ease and flexibility. 
The bad part is CSS isnt that great to scale up and split files now. with Scss variables and mixins it becomes easier to see all design code in the same component file. 


> I really don't want to drag in half of Node's ecosystem into this project if I can help it. Right now the CSS/JS bundles are built with simple UNIX and Go tools and it all works very nicely.
> 
> The only thing I want to change in the near future is to migrate from go.rice to embed

The node ecosystem is only for the developers anyways. wont be needed for the production application . 

[twtxt-pico](https://github.com/lucaslarroche/twtxt-pico) Looks great. I can try its integration with the templates. but there is a lot of Go Code in the core repo UI, that will take time to fully integrate.

**Comment by lucaslarroche** [comment]
As I'm mentioned, I take the liberty to answer.

I can't imagine that twtxt could become mainstream without a great UI.
The public pod must be must necessarily demonstrate the possibilities of twtxt. And that also a good UX.

To have a maintainable design, you must use SCSS. Preferably in the main repository.

Pico, Bootstrap or another Framework is debatable. But not SCSS. 
twtxt has to spend time on the UI/UX, and it goes through SCSS.

**Comment by prikeshsavla** [comment]
100% agree with @lucaslarroche . SCSS would be vital for UI development regardless of the CSS framework used. 
@lucaslarroche  I was gonna use gulp to build files as it give some control over how we want to build the assets with added plugins.
I see that you have used [https://prepros.io/](https://prepros.io/) . I haven't used that before. Is that really intuitive?

**Comment by lucaslarroche** [comment]
> I was gonna use gulp to build files as it give some control over how we want to build the assets with added plugins.
Gulp will be perfect. I used Prepros just because it's WYSIWIG and more easy for me.

**Comment by prikeshsavla** [comment]
So how does this sound. I give a pull request in your repo with a gulp file and browsersync config to the best of my abilities. and then we can use that as a base to build UI in the core.

**Comment by prologic** [comment]
> So how does this sound. I give a pull request in your repo with a gulp file and browsersync config to the best of my abilities. and then we can use that as a base to build UI in the core.

Doesn't it make more sense to pull in [pico-twtxt](https://github.com/lucaslarroche/twtxt-pico) as a new repo/proejct and put UI/UX efforts there?

I'm still _quite uncomfortbale_ brining into things like Gulp and a Gulpfile which itself will being in various NodeJS tooling. I _could_ compromise as long as we ensured there was a set of `Dockerfile`(s) used to build the UI. The backend as it is now already requires several C dependencies to build correctly, if you bring in additional tools, languages and their ecosystem you will also introduce a lot of complexity for very little value in my opinion. The UI/UX as it is today is quite simple and should remain that way. A more complex UI/UX is suitable as a "client" to the API with suitable Cross-Origin configuration (_not yet configurable on the backend, but could be..._).

Can we stick to incorporating Luac's changes as per my comment above, maintaining that body of work as a separate "twtxt-scss" repo with the existing styles re-incorporated back into SCSS and go from there?

We can _probably_ even embed the repo into here via Git submodule(s) (_as much as I loathe doing that due to other increased operational complexities that introduces_), or as I said earlier, just building and committing the resulting `.min.css` file as we already do today into `./internal/static/css/`.

**Comment by markwylde** [comment]
I'd like to ask what problems are we trying to solve. It seems like we've jumped into adding scss and gulp, without really defining what the problems are right now.

**Structure**
> Add Styles to the system right now is awful. Its 3 to 4 css files and the package doesnt even use a css preprocessor.

Is the issue that the css code is getting hard to manage? If so, currently the build uses minify which reduces the 4 files down to a single bundled css file. We can have as many css files in there as we want.

Maybe we break the 99-twtxt.css into separate files? We could even put them into folders.

No need to bring in a tool to do this.

**File Size**
> Also with scss we can limit the styles from Pico.css

Is the issue that the css file is too big? It's currently 10kb minified. That's pretty small.

**Variable**
> Also add our own variables and custom component styles

CSS has had [variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) now for some time. We've actually [using it currently](https://github.com/jointwt/twtxt/blob/master/internal/static/css/99-twtxt.css#L140).

Is the issue that we don't use this enough. Maybe we could use more variables instead of raw values. No scss needed here.

**Watching**
> With gulp we can also watch for *.go files with the usual html

Is the issue that it's annoying having to manually go into the terminal and run the build command? I'm not great with go, but a quick search brought up [gomon](https://github.com/c9s/gomon). Maybe that would be a better solution, as we don't need to bring in the node ecosystem to make that work.

**My thoughts**
I have used both scss, gulp and other tools that are "popular" out there, and in my opinion it just adds more overhead. A quick install of [node-sass](https://www.npmjs.com/package/node-sass) added over [200 dependencies](https://gist.github.com/markwylde/8bdb0efc8aabc22d9edd57a03129f4b1), took 15 seconds to install and had to bring in a binary as it uses a c project that building would take forever.

Again, that might be okay, it's a typical pros vs cons argument, but I'm just not seeing exactly what problems we have with development currently.

I'd prefer we create an issue for individual problems and then find a solution to them.

For example, if we want to solve the watching for changes problem, let's create a ticket "Automatically rebuild go on file change" and decide what the best solution is in there.

**Comment by prologic** [comment]
I agree with @markwylde . I'd also like to point out that the UI/UX develop experience was vastly improved with https://github.com/jointwt/twtxt/commit/4cc4fc0ab164eb028e0b48ebbe4d0149114de712 -- @prikeshsavla you brought this problem to me in our private Signal chat(s) and I fixed it there. Has this not improved the situation so you can make changes to templates and static `.css` and `.js` assets without having to rebuild? I tested changes to templates, css and js and building with `make dev`, `make build DEBUG=1` or `rm ./internal/rice-box.go && go build ./cmd/twtd/...` and running in debug mode mode with either `make dev` or `./twtd -D` works as expected. Changes to any Go templates in `./internal/templates/*.html` any any static assets in `./internal/static/*/*` are reflected in your running instance instantaneously without rebuilding or restarting the backend.

if this hasn't fixed the "reload/rebuild" problem when modifying part of the UI, then please let me know what additional problems there are and I'll be happy to fix them. Bringing in Gulp/Bower/SCSS will not in my opinion magically make this problem "solved" without adversely affecting other parts of the project and roll-on effects.

**Comment by prologic** [comment]
Also FWIW @der-lyse and @vaniot-s have recently been making UI/UX changes as well as on the Backend itself without any troubles; for example https://github.com/jointwt/twtxt/commit/d60c4aa5f762b4c0ebdd9785594d93f10a65e861 https://github.com/jointwt/twtxt/commit/469f38ff78241d1e8830b08805d2bef822065f26 and https://github.com/jointwt/twtxt/commit/7ee574c49abf8c3bf786c54a4e31266502c8a9b4

Perhaps they can share their experiences (_both good and bad_) and what other aspect we can improve.

**Comment by der-lyse** [comment]
First of all, I'm not a (Web) UI guy at all and consider myself a Unix backend programmer or something along those lines. However, I believe I'm not too terrible with CSS – but obviously can't compete with all these web hipsters out there. ;-) I personally prefer the very simple designs. I like to keep my HTML clean and to the absolute bare minimum, just semantic elements, if somehow possible. You will rarely find classes and IDs in my HTML, I like to define the CSS rules just with HTML structure, no span and div hell at all.

When working on the twtxt UI as mentioned above, I didn't have to look at the CSS itself, I just touched the HTML. That was absolutely trivial, no problems whatsoever. For this exact reply I skimmed through the CSS for the very first time and it looks okay-ish. A little bit large rule base for my liking, but I've seen a looooot worse. Reusing variables would certainly help (I didn't know they existed, thanks @markwylde!).

In the past I've used [lessc](https://pypi.org/project/lessc/), which worked quite good. The main motivation was to save typing structures over and over again and also to define variables. The latter is now directly possible within CSS itself.

I'm absolutely against introducing a new framework esp. in another language if the benefits are only very minor. @markwylde's experience above is just ridiculous and reconfirms my prejudice against tools in that particular language. This dependency hell is just terrible software engineering.

It would help to document what developers have to do exactly to work on the HTML templates/CSS/etc. It certainly would have helped me, too, that's for sure. I was too lazy to figure out the details, so `make clean dev` was probably way too much overkill. ;-) (I ended up with `make clean` when I always got a template syntax error I could not resolve, which even persisted when changing back to _master_. Then I just stuck to `clean` for good measure. Reregistering multiple times motivated me to fix the discovered label bug. ;-))

My other bad experience is with the rice box, [I complained about over there already](https://github.com/jointwt/twtxt/pull/315). But it seems that addressing it is already on the agenda.

Other than that, I'm quite pleased. I think a few lines of documentation are really the only things that are missing.

Like @prologic and @markwylde already commented, I don't know what the exact problems are to introduce such a heavy-weight change. Chances are there are much easier and simpler and thus better ways to solve them, once the particular matters are identified.

**Comment by prologic** [comment]
I agree improving developer docs are in order. I'll file a separate issue and write some 👌

**Comment by prikeshsavla** [comment]
The UI UX changes are more like fixing spaces and spelling mistakes and not
much about changing styles and creating new or updating components. I am
not undermining their efforts here, just saying it's more fixes than style
changes.

I can to try making something with plain css, but scss gives a much better
development experience. And the node sass dependency you are worrying about
is just for the developer while making style changes.

Even in Luke's repo he has used css because it is a better alternative for
splitting code and managing styles.

It's upto you what you think is good enough or not on the project.

The whole process of using gulp and browsersync is for increasing a UI
developers experience building and running the app.

Yes the new commit helped out in building the html file without the rice.go
file. But it deletes the file that is much needed for your production. You
might have to find a better way around it.

**Comment by prologic** [comment]
As I said, I agree we should improve the docs to help new developers onboard.

I still do not understand what specific problems we're attempting to solve here. Without that it really just feels like we're brining in additional tooling and overheads without even first understanding what the problems are.

Lastly even if you brought in Gulp, NPM, Browsersync, SCSS and various other tools, frameworks and their languages and ecosystems, once you leave the project and on to other things the burden is then on me. I'm not comfortable personally using tools I don't understand, don't work properly half the time or break every other time. In my experience any open-source project I come across that uses these tools (_same goes with traditional RDBMS too_), I just move on. If you can't get going in a few minutes we've failed.

That I _feel_ is my failure here right now. I've filed #319 which I'll get started on today and write some docs. Right now I don't feel adding the aforementioned dependencies into this project _really_ adds any value to the developer experience, and in fact for me would harm it as it would increase overall complexity, overheads and cognitive load.

**Comment by prologic** [comment]
Btw... One of my requirements all along for this project has been to keep things as simple and lightweight as possible. Maybe that decision won't win me a social network of billions but that's okay 😀 -- To put things in perspective, I use a piece of software called [AdGuardHome](https://github.com/AdguardTeam/AdGuardHome) also written in Go (95%) and also has a "frialy swank looking" UI. Aside from the fact that I _wouldn't have a clue_ how to contribute to any parts of that project's UI/UX (_Go not a problem!_), here's what it takes to load the UI:

<img width="1001" alt="Screen Shot 2020-12-28 at 08 25 13" src="https://user-images.githubusercontent.com/1290234/103180935-c3d31280-48e6-11eb-8dc3-51181e629f13.png">

That's:

- 16 requests
- 836kB of traffic
- 3.4MB of resources the browser has to load
- ~750ms to start interacting

Compare this with a production build of Twtxt:

<img width="1001" alt="Screen Shot 2020-12-28 at 08 28 58" src="https://user-images.githubusercontent.com/1290234/103180962-19a7ba80-48e7-11eb-9824-f4e6c7c44e92.png">

- 5 requests
- 35kB of traffic
- 102kBof resources the browser has to load
- ~35ms to start interacting

And that's reflected in the Cloudflare's stats (_which I only use to protect my home infrastructure_) (_yes I run twtxt.net from home_) where the time to first byte (TTFB) is under ~300ms (_most of that is latency to the other side of the planet!_).

Maybe I'm biased, but I strongly feel the Web is a bit broken in the sense that a) it's full of advertising (_the whole reason I wrote this thing in the first place_) b) Most of it runs poorly written, buggy client-side Javascript with all kinds of security holds, resource problems and so on and c) Most of the traffic goes to a few places (CDNs, Google, Facebook, etc).

If I had it my way, the backend would be entirely 100% HTML, no Javascript and barely any CSS at all 🤣

Don't forget there is another UI too over at [Goryon](https://github.com/jointwt/goryon) that is built on Flutter, uses the Twtxt JSON/REST APIv1 backend and uses all things Flutter :D (_This is @dooven's domain, I barely know how to build it properpy, _sigh_)

**Comment by der-lyse** [comment]
I fully agree.

> If I had it my way, the backend would be entirely 100% HTML, no Javascript and barely any CSS at all rofl

Well, let's also save those 7.6 KB and get rid of this _twtxt.min.js_, I'm totally up for it! :-) I bet the CSS can also be further reduced in size.

**Comment by prologic** [comment]
> Well, let's also save those 7.6 KB and get rid of this twtxt.min.js, I'm totally up for it! :-) I bet the CSS can also be further reduced in size.

In all seriousness the current web frontend is _supposed_ to work without JS (_degrade gracefully_); but we have to find a "good balance".

**Comment by prologic** [comment]
And it does... Tested in FF with Javascript disabled. The only thing that doesn't work is media uploads (_maybe we can have a degraded way of handling that one day..._)

**Comment by prikeshsavla** [comment]
I could careless about browsersync or npm. What I want to use is scss if you are okay with that then we can see a performant go plugin of to go for it. 

Gulp and Npm are means to an end

**Comment by markwylde** [comment]
> What I want to use is scss if you are okay with that

Sorry to disagree prikeshsavla, but I'm struggling to see how adding scss benefits the project, or what problem it solves. Please help me understand what issue we're having that SCSS would solve.

When you open one of the CSS files, what is it that you find difficult to manage. Maybe we can take it from there?

Again, why I personally think SCSS is aweful:
- I really hate inheritance/nesting. It encourages large nested blocks of code. IMO It creates a terrible development experiance.
- The mixin stuff is a "code smell" to me that your CSS is too complex. It's not needed.
- Partials and modules are unnessisary as you can split your code into files and merge them.
- Operators are unnessisary, and if you need them at runtime you can use `calc`

I know I may seem unreasonably against this here, but in my experiance as a UX developer, the SCSS codebases I have worked with have turned out to be horrendous and bloated, with long compile times. The CSS codebases I work with are small, lightweight and much easier to manage.

I try to be open to most things, but I'm not hearing any compelling reasons. Wanting to use something isn't really a good enough reason. A tool has to bring something to the table.

You wouldn't buy a chainsaw to slice some bread.

**Comment by prologic** [comment]
Sorry @prikeshsavla but @markwylde is right. There is no strong compelling reason to bring SCSS into this project. If you would still like to contribute to any of the [issues labelled with area/ui](https://github.com/jointwt/twtxt/issues?q=is%3Aissue+is%3Aopen+label%3Aarea%2Fui) you are more than welcome (_some will also require backend changes_).

Thanks all for the discussions! I also learned a lot from this 👍
